[GSoC] 237 - Typo Detection (#908)
openvinotoolkit/openvino_notebooks: notebooks/237-typo-detector/README.md

# Convert and Optimize Typo Detector with OpenVINO™

Typo detection in AI is a process of identifying and correcting typographical errors in text data using machine learning algorithms. The goal of typo detection is to improve the accuracy, readability, and usability of text by identifying and correcting mistakes made during the writing process.

A typo detector takes a sentence as an input and identify all typographical errors such as misspellings and homophone errors.

This tutorial provides how to use the [Typo Detector](https://huggingface.co/m3hrdadfi/typo-detector-distilbert-en) from the [Hugging Face Transformers](https://huggingface.co/docs/transformers/index) library to perform the above task.

The model has been pretrained on the [NeuSpell](https://github.com/neuspell/neuspell) dataset.

<img src=https://user-images.githubusercontent.com/80534358/224564463-ee686386-f846-4b2b-91af-7163586014b7.png>

</br>

> The notebook provides two methods to use the typo detector in the OpenVino runtime. I've demonstrated both methods so that you can experience both the loading of compiled models and loading models in other frameworks and converting them.

Reviewer: ```suggestion
The notebook provides two methods to run the inference of typo detector with OpenVINO runtime, so that you can experience both calling the API of Optimum with OpenVINO Runtime included, and loading models in other frameworks, converting them to OpenVINO IR format, and running inference with OpenVINO Runtime.
```

> 1. Use the [Hugging Face Optimum](https://huggingface.co/docs/optimum/index) library to load the compiled model in OpenVino IR format. Then create a pipeline with the loaded model to run inference.

Reviewer: ```suggestion
1. Use the [Hugging Face Optimum](https://huggingface.co/docs/optimum/index) library to load the compiled model in OpenVINO IR format. Then create a pipeline with the loaded model to run inference.
```
Please check to make sure "OpenVINO" is all correctly spelled in this notebook folder.

> 2. Load the model and convert to ONNX and then to OpenVino IR.

Reviewer: ```suggestion
2. Load the model and convert to ONNX and then to OpenVINO IR.
```

>    First the Pytorch model is convereted to the ONNX format and then the [Model Optimizer](https://docs.openvino.ai/latest/openvino_docs_MO_DG_Deep_Learning_Model_Optimizer_DevGuide.html) tool is used to convert to [Openvino IR format](https://docs.openvino.ai/latest/openvino_ir.html). This method provides much more insight to the openvino environment and applications.

Reviewer: ```suggestion
   First the Pytorch model is convereted to the ONNX format and then the [Model Optimizer](https://docs.openvino.ai/latest/openvino_docs_MO_DG_Deep_Learning_Model_Optimizer_DevGuide.html) tool is used to convert to [OpenVINO IR format](https://docs.openvino.ai/latest/openvino_ir.html). This method provides much more insight to how to set up a pipeline from model loading to model converting, compiling and running inference with OpenVINO, so that you could conveniently use OpenVINO to optimize and accelerate inference for other deep-learning models.
```

The following table summarises the major differences between the two methods

</br>

| Method 1                                                            | Method 2                                                           |
| ------------------------------------------------------------------- | ------------------------------------------------------------------ |
| Load models from Optimum, an extension of transformers              | Load model from transformers                                       |

> | Load the model in OpenVino IR format on the fly                     | Have to load the model and convert to ONNX and then to OpenVino IR |

Reviewer: ```suggestion
| Load the model in OpenVINO IR format on the fly                     | Convert to ONNX and then to OpenVINO IR |
```

> | Load the compiled model by default                                  | Model has to be compiled after converting to OpenVino IR           |

Reviewer: ```suggestion
| Load the compiled model by default                                  | Compile the OpenVINO IR and run inference with OpenVINO Runtime          |
```

> | In this application, loaded model from optimum is around 15% faster | The converted model is slower                                      |

Reviewer: Suggest to remove this row

| Pipeline is created to run inference with OpenVINO Runtime          | Have to manually run inference.                                    |

</br>

## Notebook Contents

This tutorial demonstrates step-by-step instructions on how to run and optimize the model.

The tutorial consists of the following steps:

- Required libraries

> - Two possible methods

Author: This part is from the "Content" section where under the two methods I have included the main content and steps of each method. It's not an explanation of the two methods. I believe this section is required as stated in the notebook guide of OpenVINO. Please let me know if I should change it.

  - Using the HuggingFace Optimum library

>     - Loading the compiled model in OpenVino IR format

Reviewer: ```suggestion
    - Loading the compiled model in OpenVINO IR format
```
